garage/tf/policies/categorical_mlp_policy_with_model.py

```diff
@@ -0,0 +1,110 @@
+"""CategoricalMLPPolicy with model."""
+from akro.tf import Discrete
+import tensorflow as tf
+
+from garage.misc.overrides import overrides
+from garage.tf.distributions import Categorical
+from garage.tf.models.discrete_mlp_model import DiscreteMLPModel
+from garage.tf.policies.base2 import StochasticPolicy2
+
+
+class CategoricalMLPPolicyWithModel(StochasticPolicy2):
+    """
+    CategoricalMLPPolicy with model.
+
+    It only works with Discrete environment.
+
+    Args:
+        env_spec: Environment specification.
+        name: variable scope of the mlp.
+        hidden_sizes: Output dimension of dense layer(s).
+        hidden_nonlinearity: Activation function for
+                    intermediate dense layer(s).
+        output_nonlinearity: Activation function for
+                    output dense layer.
+        layer_normalization: Bool for using layer normalization or not.
+
+    """
+
+    def __init__(self,
+                 env_spec,
+                 name="CategoricalMLPPolicy",
+                 hidden_sizes=(32, 32),
+                 hidden_nonlinearity=tf.nn.tanh,
+                 output_nonlinearity=tf.nn.softmax,
+                 layer_normalization=False):
+        assert isinstance(env_spec.action_space, Discrete), (
+            "CategoricalMLPPolicy only works with Discrete environment.")
+        super().__init__(name, env_spec)
+        self.obs_dim = env_spec.observation_space.flat_dim
+        self.action_dim = env_spec.action_space.n
+
+        self.model = DiscreteMLPModel(
+            output_dim=self.action_dim,
+            name=name,
+            hidden_sizes=hidden_sizes,
+            hidden_nonlinearity=hidden_nonlinearity,
+            output_nonlinearity=output_nonlinearity,
+            layer_normalization=layer_normalization)
+
+        self._initialize()
+
+    def _initialize(self):
+        state_input = tf.placeholder(tf.float32, shape=(None, self.obs_dim))
+
+        self.model.build(state_input)
+
+        self._f_prob = tf.get_default_session().make_callable(
+            self.model.networks['default'].output,
+            feed_list=[self.model.networks['default'].input])
+
+    @property
+    def vectorized(self):
+        """Vectorized or not."""
+        return True
+
+    @overrides
+    def dist_info_sym(self, obs_var, state_info_vars=None, name=None):
+        """Symbolic graph of the distribution."""
+        prob = self.model.build(obs_var, name=name)[0]
+        return dict(prob=prob)
+
+    @overrides
+    def dist_info(self, obs, state_infos=None):
+        """Distribution info."""
+        return dict(prob=self._f_prob(obs))
+
+    # The return value is a pair. The first item is a matrix (N, A), where each
```

Comment:
Since we are refactoring I think we should delete this incorrect comment. Or change it as the second item is also (N, A). Please correct me if I'm wrong.